pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go

```diff
@@ -0,0 +1,282 @@
+				new := n.(*extensionsv1alpha1.Worker)
+				return validation.ValidateWorker(new)
+			},
+			validateUpdateResource: func(n, o client.Object) field.ErrorList {
+				new := n.(*extensionsv1alpha1.Worker)
+				old := o.(*extensionsv1alpha1.Worker)
+				return validation.ValidateWorkerUpdate(new, old)
+			}})
+
+	return &h
+}
+
+type handler struct {
+	decoder   *admission.Decoder
+	logger    logr.Logger
+	artifacts map[metav1.GroupVersionResource]artifact
+}
+
+func (h *handler) addResource(gvr metav1.GroupVersionResource, art artifact) {
+	h.artifacts[gvr] = art
+}
+
+var _ admission.Handler = &handler{}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *handler) Handle(ctx context.Context, ar admission.Request) admission.Response {
+	h.logger.Info(fmt.Sprintf("validating resource of type %s for operation %s", ar.Resource, ar.Operation))
+
+	_, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	artifact, ok := h.artifacts[ar.Resource]
+	if !ok {
+		return admission.Allowed("validation not found for the given resource")
+	}
+
+	switch ar.Operation {
+	case admissionv1.Create:
+		return h.handleValidation(ar.Object, ar.OldObject, artifact.newEntity, artifact.validateCreateResource)
+	case admissionv1.Update:
+		return h.handleValidation(ar.Object, ar.OldObject, artifact.newEntity, artifact.validateUpdateResource)
+	default:
+		return admission.Allowed("operation is not CREATE or UPDATE")
+	}
+}
+
+type newEntity func() client.Object
+type validate func(new, old client.Object) field.ErrorList
+
+// artifact servers as a helper to setup the corresponding function.
+type artifact struct {
+	// newEntity is a simple function that creates and returns a new resource.
+	newEntity newEntity
+
+	// validateCreateResource is a wrapper function for the different create validation functions.
+	validateCreateResource validate
+
+	// validateUpdateResource is a wrapper function for the different update validation functions.
+	validateUpdateResource validate
+}
+
+func (h handler) handleValidation(object, oldObject runtime.RawExtension, newEntity newEntity, validate validate) admission.Response {
+	obj := newEntity()
+	if err := h.decoder.DecodeRaw(object, obj); err != nil {
+		h.logger.Error(err, "could not decode object", "object", object)
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("could not decode object %v: %w", object, err))
+	}
+
+	h.logger.Info(fmt.Sprintf("handle validation for %s", obj.GetName()))
+
+	oldObj := newEntity()
+	if len(oldObject.Raw) != 0 {
+		if err := h.decoder.DecodeRaw(oldObject, oldObj); err != nil {
+			h.logger.Error(err, "could not decode old object", "old object", oldObj)
+			return admission.Errored(http.StatusBadRequest, fmt.Errorf("could not decode old object %v: %v", oldObj, err))
+		}
+	}
+
+	errors := validate(obj, oldObj)
+	if len(errors) != 0 {
+		err := apierrors.NewInvalid(gvk.GroupKind(), obj.GetName(), errors)
```

Comment:
This hardcoded `gvk` here must not be used. Instead, you should construct the GVK of the object that is failing the validation. You should already have this information in `obj` and the GV part also in the key of the `artifacts` map you have constructed previously.